extension/src/popup/components/account/AccountAssets/index.tsx

```diff
@@ -21,25 +23,40 @@ export const AssetIcon = ({
   code: string;
   issuerKey: string;
   retryAssetIconFetch?: (arg: { key: string; code: string }) => void;
-}) =>
-  assetIcons[getCanonicalFromAsset(code, issuerKey)] || code === "XLM" ? (
-    <img
-      className="AccountAssets__asset--logo"
-      alt={`${code} logo`}
-      src={
-        code === "XLM"
-          ? StellarLogo
-          : assetIcons[getCanonicalFromAsset(code, issuerKey)] || ""
-      }
-      onError={() => {
-        if (retryAssetIconFetch) {
-          retryAssetIconFetch({ key: issuerKey, code });
-        }
-      }}
-    />
+}) => {
+  const isXlm = code === "XLM";
+  const [hasError, setHasError] = useState(false);
+  const [isLoading, setIsLoading] = useState(!isXlm);
+  const canonicalAsset = assetIcons[getCanonicalFromAsset(code, issuerKey)];
+  const imgSrc = hasError ? ImageMissingIcon : canonicalAsset || "";
+
+  return canonicalAsset || isXlm ? (
+    <div
+      className={`AccountAssets__asset--logo ${
+        hasError ? "AccountAssets__asset--error" : ""
+      } ${isLoading ? "AccountAssets__asset--loading" : ""}`}
+    >
+      <img
+        alt={`${code} logo`}
+        src={isXlm ? StellarLogo : imgSrc}
+        onError={() => {
+          if (retryAssetIconFetch) {
+            retryAssetIconFetch({ key: issuerKey, code });
```

Comment:
❓ definitely an edge case, but if the retry function fails, then `hasError` will stay as false right?

not sure what the best fix would be 🤔  we could have retryAssetIconFetch return something that could be used to check if failed/succeeded

Reply:
Good catch - `setHasError(true)` should always be called. If retry works, it'll re-render and skip right over onError to onLoad

`setIsLoading(false)` actually doesn't really need to be called onError bc if the icon is never found, it just skips down to line 54 where isLoading doesn't have any effect